### PR TITLE
fileinput first version

### PR DESCRIPTION
**Pull request by jukebox**

### CODE REVIEW

File: stdlib/3/fileinput.pyi [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from typing import Iterable, BinaryIO
+
+__all__ = ...   # type: List[str]
+_state  = ...   # type: None
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: We don't need unexported, undocumented variables.

File: stdlib/3/fileinput.pyi [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+def input(
+    files               = None,
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: What's the type? I presume `List[something]`?

File: stdlib/3/fileinput.pyi [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+__all__ = ...   # type: List[str]
+_state  = ...   # type: None
+
+def input(
+    files               = None, 
+    inplace:    bool    = ...,
+    backup:     str     = ...,
+    bufsize:    int     = ..., 
+    mode:       str     = ...,
+    openhook            = None): ...
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: I guess the type should be `Callable[[something], something_else]`?

File: stdlib/3/fileinput.pyi [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        inplace:    bool    = ...,
+        backup:     str     = ...,
+        bufsize:    int     = ...,
+        mode:       str     = ...,
+        openhook            = None) -> None : ...
+
+
+    def __del__(self)                           -> None : ...
+    def close(self)                             -> None : ...
+    def __enter__(self)                                 : ... #return self
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: Why the `#return` comments? I think these aren't needed or useful.

File: stdlib/3/fileinput.pyi [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    def readline(self)                          -> str  : ...
+    def filename(self)                          -> str  : ...
+    def lineno(self)                            -> int  : ...
+    def filelineno(self)                        -> int  : ...
+    def fileno(self)                            -> int  : ...
+    def isfirstline(self)                       -> bool : ...
+    def isstdin(self)                           -> bool : ...
+
+def hook_compressed(filename: str, mode: str)   -> BinaryIO : ...
+def hook_encoded(encoding: str)                 -> BinaryIO : ...
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: General remark: PEP 8 doesn't want you to line up the arrows, colons etc. Please follow that style.

Comment (jukebox) [inline stdlib/3/fileinput.pyi]: ok i will fix that, . maybe you can check pep8 in travis ?

File: stdlib/3/fileinput.pyi (hunk 4 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: fixed

**Comment by gvanrossum** [comment]
If you need guidance you can easily run pep8 yourself. It would be a massive undertaking to make all of typeshed conform to all of pep8's picky rules, and some of the rules don't apply (e.g. about blank lines between methods).

File: stdlib/3/fileinput.pyi (hunk 1 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: ok, done.

File: stdlib/3/fileinput.pyi (hunk 5 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: done

File: stdlib/3/fileinput.pyi (hunk 2 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: done

File: stdlib/3/fileinput.pyi (hunk 3 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: fixed

**Comment by jukebox** [comment]
new version

File: stdlib/3/fileinput.pyi [hunk 6]
```diff
@@ -0,0 +1,49 @@
+from typing import Iterable, BinaryIO, List
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: The tests fail because you're missing `Callable` here.

Comment (jukebox) [inline stdlib/3/fileinput.pyi]: yes sorry i fixed this

File: stdlib/3/fileinput.pyi [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def input(
+    files: List[str]=...,
+    inplace: bool=...,
+    backup: str=...,
+    bufsize: int=...,
+    mode: str=...,
+    openhook: Callable[[str, str], BinaryIO]=...
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: I think all occurrences of BinaryIO need to be changed to IO[AnyStr]. The module appears to work for binary as well as text files, depending on the value of the mode argument. The return value of `__next__()` and other methods that return data from the file need to be AnyStr as well. The PEP 484 type system currently cannot express this properly. I think the FileInput class needs to be explicitly inheriting from Iterable[AnyStr] too, and you'd have to add Generic[AnyStr] too I think (please write some test files yourself to verify these edge cases).

Comment (jukebox) [inline stdlib/3/fileinput.pyi]: done. Why FileInput doest inherit from collection.abc.Iterable ?

File: stdlib/3/fileinput.pyi [hunk 8]
```diff
@@ -0,0 +1,51 @@
+from typing import Iterable, Callable, IO, AnyStr, Generic, Any, TypeVar
+
+StrOrIterableStr = TypeVar('StrOrIterableStr', str, Iterable[str])
```
Comment (gvanrossum) [inline stdlib/3/fileinput.pyi]: I think this should be Union[str, Iterable[str]] -- no TypeVar needed.

**Comment by gvanrossum** [comment]
Hi, this is getting better all the time! I wrote a little test program myself and tried several variants:
```
import fileinput

f = fileinput.FileInput(mode='r')  # type: fileinput.FileInput[str]
for line in f:
    if f.isfirstline():
        print("NEW FILE:", f.filename())
    print("%s:%d: %s" % (f.filename(),
                         f.lineno() ,
                         line.rstrip('\r\n')))
```
I varied this a bit, e.g. switch the mode to 'rb' and the type decl to types; I also wrote a variant that uses the module-level functions. I ended up making a few improvements, which I am posting here:
```
diff --git a/stdlib/3/fileinput.pyi b/stdlib/3/fileinput.pyi
index 3f01792..1eaf8a7 100644
--- a/stdlib/3/fileinput.pyi
+++ b/stdlib/3/fileinput.pyi
@@ -1,6 +1,6 @@
-from typing import Iterable, Callable, IO, AnyStr, Generic, Any, TypeVar
+from typing import Iterable, Iterator, Callable, IO, AnyStr, Generic, Any, Union
 
-StrOrIterableStr = TypeVar('StrOrIterableStr', str, Iterable[str])
+StrOrIterableStr = Union[str, Iterable[str]]
 
 
 def input(
@@ -10,7 +10,7 @@ def input(
     bufsize: int=...,
     mode: str=...,
     openhook: Callable[[str, str], IO[AnyStr]]=...
-    ): ...
+    ) -> Iterable[AnyStr]: ...
 
 
 def close() -> None: ...
@@ -33,13 +33,13 @@ class FileInput(Iterable[AnyStr], Generic[AnyStr]):
 
     def __del__(self) -> None: ...
     def close(self) -> None: ...
-    def __enter__(self): ...
+    def __enter__(self) -> 'FileInput[AnyStr]': ...
     def __exit__(self, type: Any, value: Any, traceback: Any) -> None: ...
-    def __iter__(self): ...
-    def __next__(self) -> str: ...
-    def __getitem__(self, i) -> str: ...
+    def __iter__(self) -> Iterator[AnyStr]: ...
+    def __next__(self) -> AnyStr: ...
+    def __getitem__(self, i) -> AnyStr: ...
     def nextfile(self) -> None: ...
-    def readline(self) -> str: ...
+    def readline(self) -> AnyStr: ...
     def filename(self) -> str: ...
     def lineno(self) -> int: ...
     def filelineno(self) -> int: ...
```
Let me know if you want to apply those or if you want me to just merge the PR and apply these fixes myself.

File: stdlib/3/fileinput.pyi (hunk 8 above)
Comment (jukebox) [inline stdlib/3/fileinput.pyi]: fixed, sorry

**Comment by gvanrossum** [comment]
Thanks -- no worries! It's your first contribution here and you're doing great. Thanks for working through this. I'm merging it now. Do you have the energy to create the Python 2 stub too?
